pkg/apis/networking/validation/validation.go

```diff
@@ -808,3 +808,65 @@ func ValidateIPAddressUpdate(update, old *networking.IPAddress) field.ErrorList
 	allErrs = append(allErrs, apivalidation.ValidateImmutableField(update.Spec.ParentRef, old.Spec.ParentRef, field.NewPath("spec").Child("parentRef"))...)
 	return allErrs
 }
+
+var ValidateServiceCIDRName = apimachineryvalidation.NameIsDNSSubdomain
+
+func ValidateServiceCIDR(cidrConfig *networking.ServiceCIDR) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&cidrConfig.ObjectMeta, false, ValidateServiceCIDRName, field.NewPath("metadata"))
+
+	if cidrConfig.Spec.IPv4 == "" && cidrConfig.Spec.IPv6 == "" {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("Spec"), cidrConfig.Spec, "at least one CIDR required"))
+	}
+
+	if cidrConfig.Spec.IPv4 != "" {
+		prefix, err := netip.ParsePrefix(cidrConfig.Spec.IPv4)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("IPv4"), cidrConfig.Spec.IPv4, err.Error()))
+		} else {
+			if prefix.Addr() != prefix.Masked().Addr() {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv4"), cidrConfig.Spec.IPv4, "wrong CIDR format, IP doesn't match network IP address"))
+			}
+			if prefix.Masked().String() != cidrConfig.Spec.IPv4 {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv4"), cidrConfig.Spec.IPv4, "CIDR not in canonical format"))
+			}
+			if !prefix.Addr().Is4() {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv4"), cidrConfig.Spec.IPv4, "not IPv4 family CIDR"))
+			}
+		}
+	}
+
+	if cidrConfig.Spec.IPv6 != "" {
+		prefix, err := netip.ParsePrefix(cidrConfig.Spec.IPv6)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("IPv6"), cidrConfig.Spec.IPv6, err.Error()))
+		} else {
+			if prefix.Addr() != prefix.Masked().Addr() {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv6"), cidrConfig.Spec.IPv6, "wrong CIDR format, IP doesn't match network IP address"))
+			}
+			if prefix.Masked().String() != cidrConfig.Spec.IPv6 {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv6"), cidrConfig.Spec.IPv6, "CIDR not in canonical format"))
+			}
+			if !prefix.Addr().Is6() {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv6"), cidrConfig.Spec.IPv6, "not IPv6 family CIDR"))
+			}
+		}
+	}
+
+	return allErrs
```

Comment:
Oh, I was thinking you could use a /32 but I guess that wouldn't work since it won't assign the first or last IP in a range... though we could special-case /32 and /128...

Anyway, the idea was that you could start the apiserver with a `--service-cluster-ip-range` that contained only a single IP, which would then get assigned to kubernetes.default, and then once it starts, you would create a ServiceCIDR which would define the CIDR to be used by every other Service.

This was an attempt to minimize the ugliness of the fact that Service CIDR configuration is now partially on the command-line and partially in API objects, by only using the command-line to configure the 1 Service IP that *can't* be configured by an API object. Although actually, that's not true anyway; you can connect to the apiserver via the node IP even before the service network is working. So I guess the better version of this proposal would be to allow starting kube-apiserver up without passing `--service-cluster-ip-range` at all, and it would just adopt the first ServiceCIDR it sees as the default. Or something?

Reply:
we have to keep the same validation we have already for the flag, that AFAIK only checks the max size https://github.com/kubernetes/kubernetes/blob/69c3b23abdbda53d14e14afc2af2bdfef23ac7b0/cmd/kube-apiserver/app/options/validation.go#L36, with this feature we remove that restriction, that is something that I realized that I missed